pkg/agent/util/net_linux.go

```
@@ -332,7 +332,12 @@ func ConfigureLinkRoutes(link netlink.Link, routes []interface{}) error {
 	for _, r := range routes {
 		rt := r.(netlink.Route)
 		rt.LinkIndex = netlinkAttrs(link).Index
-		if err := netlinkUtil.RouteReplace(&rt); err != nil {
+		if err := netlinkUtil.RouteAdd(&rt); err != nil {
```

Functionality-wise, I think the change is ok. However, I still don't see a _legitimate_ case for `RouteAdd` failing, so it feels that the log message should be an error to be easily identified?

Going back to the original issue, I can see how we could have 2 "default" routes with different metrics. In that case, the `RouteReplace` should behave as expected, as the `metric` should be taken into account when matching. I cannot even create 2 "default" routes if I don't use different `metric` values, even if the routes are for different interfaces...

It may be slightly better to do `RouteDelete` followed by a `RouteAdd` IMO. It will not suffer from the same issue as `RouteReplace` because `RouteDelete` should take into account all the optional attributes, including the outgoing interface (`LinkIndex`). We would ignore any error returned by `RouteDelete`, and log an error if `RouteAdd` fails. What do you think @tnqn @luolanzone?

---

Hi @antoninbas when there are two interfaces are configured via `systemd-networkd`, e.g.

```
 cat 10-cloud-init-eth0.network
[Address]
Address=172.26.0.3/27

[Match]
MACAddress=04:50:56:00:f4:00
Name=eth0

[Network]
DHCP=no
DNS=192.19.189.10

[Route]
Gateway=172.26.0.1
```
```
cat 10-cloud-init-eth1.network
[Address]
Address=172.26.2.3/24

[Match]
MACAddress=04:50:56:00:9c:00
Name=eth1

[Network]
DHCP=no
DNS=192.19.189.10

[Route]
Gateway=172.26.2.1
```

You will get two routes like following even-though we can't add such routes manually.
```
default via 172.26.0.1 dev eth0 proto static
default via 172.26.2.1 dev eth1 proto static
```

It means if we delete the route first, then it will delete the route (e.g `default via 172.26.2.1 dev eth1 proto static`) which are automatically configured by `systemd-networkd` when the eth1 is up again. But the `RouteAdd` will still fail because of another default route still exists, e.g `default via 172.26.0.1 dev eth0 proto static`  in this example.  It seems when users run 'ip route add' command in a terminal, it will check both destination ( `default` ) and proto ( `proto static`) to disallow users to add such route. 

The Add command will check the 'destination' and 'proto' or 'metric',  which means when there is already a route like `default via 192.168.77.1 dev eth1 metric 100`, a new route like `default via 172.18.0.1 dev eth0 metric 100` will also fail via `ip route add`. But for systemd-networkd daemon, it can add two routes with `proto static` or `metric 100`.

The above routes should not be a typical case,  I believe it may even cause issue as well because the route might be randomly chose.  So the question is, do we want to delete those routes automatically added back by network daemon?

---

So I finally was able to understand how this can happen. You can reproduce the same situation using `ip route append` instead of `ip route add`. In this case, you can end up with 2 "different" default routes with the same metric. This effectively translates into a multipath route, and flows can be load-balanced.
Still this is probably not the desired behavior here, and the routes should really be configured with different metrics, or the second interface should be configured *not* to have a default route.

I am not sure we can realistically support the multipath route case (if there is no network manager for example). That would make the code extra complex.